packages/cw-pre-propose-base/src/execute.rs

```diff
@@ -0,0 +1,320 @@
+        if info.sender != dao {
+            Err(PreProposeError::NotDao {})
+        } else {
+            let denom = match denom {
+                Some(denom) => Some(denom.into_checked(deps)?),
+                None => {
+                    let config = self.config.load(deps.storage)?;
+                    config.deposit_info.map(|d| d.denom)
+                }
+            };
+            match denom {
+                None => Err(PreProposeError::NoWithdrawalDenom {}),
+                Some(denom) => {
+                    let balance = denom.query_balance(&deps.querier, &env.contract.address)?;
+                    if balance.is_zero() {
+                        Err(PreProposeError::NothingToWithdraw {})
+                    } else {
+                        let withdraw_message = denom.get_transfer_to_message(&dao, balance)?;
+                        Ok(Response::default()
+                            .add_message(withdraw_message)
+                            .add_attribute("method", "withdraw")
+                            .add_attribute("receiver", &dao)
+                            .add_attribute("denom", denom.to_string()))
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn execute_status_change_proposal_hook(
+        &self,
+        deps: Deps,
+        info: MessageInfo,
+        id: u64,
+        new_status: String,
+    ) -> Result<Response, PreProposeError> {
+        let proposal_module = self.proposal_module.load(deps.storage)?;
+        if info.sender != proposal_module {
+            return Err(PreProposeError::NotModule {});
+        }
+
+        match self.deposits.may_load(deps.storage, id)? {
+            Some((deposit_info, proposer)) => {
+                let messages = if let Some(ref deposit_info) = deposit_info {
+                    // If the proposal is completed, either return to the DAO
+                    // or issue a refund.
+                    let proposal_completed = new_status == "closed" || new_status == "executed";
+
+                    if proposal_completed {
+                        // Refund can be issued if proposal if it is going to
+                        // closed or executed.
+                        let should_refund_to_proposer = (new_status == "closed"
+                            && deposit_info.refund_policy == DepositRefundPolicy::Always)
+                            || (new_status == "executed"
+                                && deposit_info.refund_policy != DepositRefundPolicy::Never);
+
+                        if should_refund_to_proposer {
+                            deposit_info.get_return_deposit_message(&proposer)?
+                        } else {
+                            // If the proposer doesn't get the deposit, the DAO does.
+                            let dao = self.dao.load(deps.storage)?;
+                            deposit_info.get_return_deposit_message(&dao)?
+                        }
+                    } else {
```

Comment:
yeah upon further reflection about the hooks intention, since we only care about initial and final status, I think it would be nicer to stop using the generic proposal hooks and only call this hook explicitly on `execute_execute` and `execute_close`. as we discussed earlier using the proposal hooks creates some unnecessary custom logic

Reply:
this is such a good idea. perhaps all this module needs is two hooks `ProposalCreatedHook { id: u64 }` and `ProposalCompletedHook { id: u64, passed: bool }`?